protocols/identify/src/identify.rs

```diff
@@ -0,0 +1,177 @@
+    pub fn new(protocol_version: String, agent_version: String) -> Self {
+        Identify {
+            protocol_version,
+            agent_version,
+            observed_addresses: HashMap::new(),
+            to_answer: SmallVec::new(),
+            futures: SmallVec::new(),
+            events: VecDeque::new(),
+        }
+    }
+}
+
+impl<TSubstream, TTopology> NetworkBehaviour<TTopology> for Identify<TSubstream>
+where
+    TSubstream: AsyncRead + AsyncWrite,
+{
+    type ProtocolsHandler = ProtocolsHandlerSelect<IdentifyListenHandler<TSubstream>, PeriodicIdentification<TSubstream>>;
+    type OutEvent = IdentifyEvent;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        IdentifyListenHandler::new().select(PeriodicIdentification::new())
+    }
+
+    fn inject_connected(&mut self, peer_id: PeerId, endpoint: ConnectedPoint) {
+        let observed = match endpoint {
+            ConnectedPoint::Dialer { address } => address,
+            ConnectedPoint::Listener { send_back_addr, .. } => send_back_addr,
+        };
+
+        self.observed_addresses.insert(peer_id, observed);
+    }
+
+    fn inject_disconnected(&mut self, peer_id: &PeerId, _: ConnectedPoint) {
+        self.observed_addresses.remove(peer_id);
+    }
+
+    fn inject_node_event(
+        &mut self,
+        peer_id: PeerId,
+        event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
+    ) {
+        match event {
+            EitherOutput::Second(PeriodicIdentificationEvent::Identified(remote)) => {
+                self.events
+                    .push_back(NetworkBehaviourAction::GenerateEvent(IdentifyEvent::Identified {
+                        peer_id: peer_id,
+                        info: remote.info,
+                        observed_addr: remote.observed_addr.clone(),
+                    }));
+                self.events
+                    .push_back(NetworkBehaviourAction::ReportObservedAddr {
+                        address: remote.observed_addr,
+                    });
+            }
+            EitherOutput::First(sender) => {
+                let observed = self.observed_addresses.get(&peer_id)
+                    .expect("We only receive events from nodes we're connected to ; we insert into the \
+                            hashmap when we connect to a node and remove only when we disconnect; QED");
+                self.to_answer.push((sender, observed.clone()));
+            }
+            _ => (), // TODO: exhaustive pattern
+        }
+    }
+
+    fn poll(
+        &mut self,
+        params: &mut PollParameters<TTopology>,
+    ) -> Async<
+        NetworkBehaviourAction<
+            <Self::ProtocolsHandler as ProtocolsHandler>::InEvent,
+            Self::OutEvent,
+        >,
+    > {
+        if let Some(event) = self.events.pop_front() {
+            return Async::Ready(event);
+        }
+
+        while !self.to_answer.is_empty() {
+            let (sender, observed) = self.to_answer.remove(0);
```

Comment:
Why not 
```rust
        for (sender, observed) in self.to_answer.drain() {
```
instead of
```rust
        while !self.to_answer.is_empty() {
            let (sender, observed) = self.to_answer.remove(0);
```
?